lib/index.js

```diff
@@ -59,6 +59,7 @@ function AssetGraph(options) {
 
     // this.root might be undefined, in which case urlTools.urlOrFsPathToUrl will use process.cwd()
     this.root = normalizeUrl(urlTools.urlOrFsPathToUrl(this.root, true)); // ensureTrailingSlash
+    this.canonicalRoot = normalizeUrl(urlTools.urlOrFsPathToUrl(this.canonicalRoot, true)); // ensureTrailingSlash
```

Comment:
We should probably require `canonicalRoot` to be a url, so `urlTools.urlOrFsPathToUrl` won't help.

Hmm, we could also allow specifying the canonical root as a source expression (https://www.w3.org/TR/CSP3/#grammardef-source-expression, our impl. here: https://github.com/assetgraph/assetgraph/blob/master/lib/matchSourceExpression.js) so you can also express the "both http and https" case. That might help us get rid of some of the special casing of protocol-relative urls when we're populating from `file://`.